app/controllers/api/v3/courses_controller.rb

```diff
@@ -29,10 +40,35 @@ def fields_param
 
     private
 
+      def next_page
+        goto_page(@courses.next_page) if @courses.next_page.present?
+      end
+
+      def previous_page
+        goto_page(@courses.prev_page) if @courses.prev_page.present?
+      end
+
+      def goto_page(index)
+        params.then do |new_params|
+          new_params[:page][:page] = index
+          url_for(new_params.to_unsafe_h.merge(only_path: true))
+        end
+      end
+
+      def build_courses
+        @courses = if @provider.present?
+                     @provider.courses
+                   else
+                     @recruitment_cycle.courses
+                   end
+      end
```

Comment:
Looking at the server logs this appears to be an N+1 query loading `FinancialIncentive` and `CourseEnrichment` separately.

Reply:
This is now fixed, FinancialIncentives just needed an include in a service, while course enrichments was fixed by simply excluding unnecessary fields.